src/main/java/com/lovetropics/minigames/common/core/game/behavior/action/GameActionList.java

```diff
@@ -13,92 +12,82 @@
-import net.minecraft.util.StringRepresentable;
 
 import java.util.Arrays;
 import java.util.List;
 import java.util.function.Function;
+import java.util.function.Supplier;
+
+public class GameActionList<T, A extends ActionTarget<T>> {
+    public static final GameActionList<ServerPlayer, PlayerActionTarget> EMPTY = new GameActionList<>(List.of(), PlayerActionTarget.SOURCE);
+
+    public static final MapCodec<GameActionList<?, ?>> MAP_CODEC = RecordCodecBuilder.mapCodec(i -> i.group(
+            MoreCodecs.listOrUnit(IGameBehavior.CODEC).fieldOf("actions").forGetter(list -> list.behaviors),
+            ActionTarget.FALLBACK_PLAYER.optionalFieldOf("target", PlayerActionTarget.SOURCE).forGetter(list -> list.target)
+    ).apply(i, GameActionList::new));
+
+    private static final Codec<GameActionList<?, ?>> SIMPLE_CODEC = MoreCodecs.listOrUnit(IGameBehavior.CODEC)
+            .flatComapMap(
+                    behaviors -> new GameActionList<>(behaviors, PlayerActionTarget.SOURCE),
+                    list -> {
+                        if (!(list.target instanceof PlayerActionTarget tg) || tg.target() != PlayerActionTarget.Target.SOURCE) {
+                            return DataResult.error(() -> "Cannot encode simple action list with target: " + list.target);
+                        }
+                        return DataResult.success(list.behaviors);
+                    }
+            );
+
+    public static final Codec<GameActionList> CODEC = Codec.either(SIMPLE_CODEC, MAP_CODEC.codec())
+            .xmap(either -> either.map(Function.identity(), Function.identity()), Either::right);
+    public static final Codec<GameActionList<?, ?>> TYPE_SAFE_CODEC = (Codec<GameActionList<?,?>>) (Object) CODEC;
+
+    public static <T, A extends ActionTarget<T>> Codec<GameActionList<T, A>> mandateType(Supplier<Codec<A>> type) {
+        final Function<GameActionList<?, ?>, DataResult<GameActionList<T, A>>> function = gameActionList -> {
+            if (gameActionList.target.type() != type.get()) {
+                return DataResult.error(() -> "Action list target must be of type: " + ActionTargetTypes.REGISTRY.get().getKey(type.get()));
+            }
+            return DataResult.success((GameActionList<T, A>) gameActionList);
+        };
+        return TYPE_SAFE_CODEC.flatXmap(function, function);
+    }
+
+    private final List<IGameBehavior> behaviors;
+    public final A target;
+
+    private final GameEventListeners listeners = new GameEventListeners();
+
+    public GameActionList(List<IGameBehavior> behaviors, A target) {
+        this.behaviors = behaviors;
+        this.target = target;
+    }
+
+    public void register(IGamePhase game, EventRegistrar events) {
+        for (IGameBehavior behavior : behaviors) {
+            behavior.register(game, events.redirect(GameActionEvents::matches, listeners));
+        }
+    }
+
+    public boolean applyPlayer(IGamePhase game, GameActionContext context, ServerPlayer... sources) {
```

Comment:
These can probably all fall away if we're strict about `T` when we define `GameActionList`